database/src/service/read.rs

```diff
@@ -47,145 +49,163 @@ type ResponseSend = tokio::sync::oneshot::Sender<ResponseResult>;
 /// Calling [`tower::Service::call`] with a [`DatabaseReadHandle`] & [`ReadRequest`]
 /// will return an `async`hronous channel that can be `.await`ed upon
 /// to receive the corresponding [`Response`].
-#[derive(Clone, Debug)]
+#[derive(Clone)]
 pub struct DatabaseReadHandle {
-    /// Sender channel to the database read thread-pool.
+    /// Handle to the custom `rayon` DB reader thread-pool.
+    ///
+    /// Requests are "spawn"ed in this thread-pool, and responses
+    /// are returned via a channel we (the caller) provide.
+    pub(super) pool: Arc<rayon::ThreadPool>,
+
+    /// Access to the database.
+    pub(super) db: Arc<ConcreteEnv>,
+}
+
+impl DatabaseReadHandle {
+    /// Initialize the `DatabaseReader` thread-pool backed by `rayon`.
     ///
-    /// We provide the response channel for the thread-pool.
-    pub(super) sender: crossbeam::channel::Sender<(ReadRequest, ResponseSend)>,
+    /// This spawns `N` amount of `DatabaseReader`'s
+    /// attached to `db` and returns a handle to the pool.
+    ///
+    /// Should be called _once_ per actual database.
+    #[cold]
+    #[inline(never)] // Only called once.
+    pub(super) fn init(db: &Arc<ConcreteEnv>, reader_threads: ReaderThreads) -> Self {
+        // How many reader threads to spawn?
+        let reader_count = reader_threads.as_threads();
+
+        // Spawn `rayon` reader threadpool.
+        let pool = rayon::ThreadPoolBuilder::new()
+            .num_threads(reader_count.get())
+            .thread_name(|i| format!("cuprate_helper::service::read::DatabaseReader{i}"))
+            .build()
+            .unwrap();
+
+        // Return a handle to the pool.
+        Self {
+            pool: Arc::new(pool),
+            db: Arc::clone(db),
+        }
+    }
+
+    /// TODO
+    #[inline]
+    pub const fn db(&self) -> &Arc<ConcreteEnv> {
+        &self.db
+    }
 }
 
 impl tower::Service<ReadRequest> for DatabaseReadHandle {
     type Response = Response;
     type Error = RuntimeError;
-    type Future = ResponseRecv;
+    type Future = ResponseReceiver;
 
     #[inline]
     fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        todo!()
+        // Threadpool is always ready as long as this handle is alive.
+        Poll::Ready(Ok(()))
     }
```

Comment:
Should we do something here when the threadpool is at capacity?

Reply:
Yeah we do need some sort of backpressure here to make sure we don't get overloaded with requests.

I think the best way to do this is to have a tokio semaphore and when `poll_ready` is called we acquire a permit, when the permit is acquired we return Poll::Ready.

We then pass this permit into the rayon pool so when the task is completed the permit is dropped.

The amount of permits should probably be the amount of threads we have in the pool

Reply:
Where would we store the `SemaphorePermit<'_>` in-between `poll_ready()` and `call()`?

Reply:
Ah, like this?
```rust
pub struct DatabaseReadHandle {
    pool: Arc<rayon::ThreadPool>,
    semaphore: Arc<Semaphore>,
    permit: Option<OwnedSemaphorePermit>,
    env: Arc<ConcreteEnv>,
}
```
Would it be something like this in `call()`?
```rust
let permit = selt.permit
    .take()
    .expect("we should have acquired a permit in poll_ready()");

/* pass to rayon pool */
```

Reply:
Yep

> Although OwnedSemaphorePermit is !Clone so we'd have to manually impl and set it as None.

That's ok

Reply:
you will need this though: https://docs.rs/tokio-util/latest/tokio_util/sync/struct.PollSemaphore.html#